### ISSUE TITLE
Insert between paragraphs ("at cursor" vs "after the active block")

### ISSUE THREAD

**Issue by folletto** [WordPress/gutenberg#1710]
![screen shot 2017-07-04 at 17 43 13](https://user-images.githubusercontent.com/4389/27838104-4b7a0a60-60e0-11e7-99bf-e649746c1e63.png)

I suggest to allow the "Insert" button from the top bar to insert always where the cursor is instead that strictly at the end of the block — at least for text blocks.

### Suggested Behaviour

1. Write a post, multiple paragraphs.
2. I want to add an image (or other block) between two paragraphs.
3. I click on "Insert" and the new block gets added in the middle, splitting the text block in two blocks.

I'm aware there are a lot of considerations to make here, so it's just a conversation starter. To name a few:

* If there's a cursor, is the cursor the insert point? I'd say yes, it's expected, when a block that has a cursor is selected.
* Do we allow this only if the cursor is on a new line, or always? (probably always, but there are pros and cons).
* Should we then provide an API for blocks that if they implemented, then it allows splitting at cursor? (i.e. the blocks might return two instances of self with the split done at the right point).

### Extended Description

Today I was trying a bit further Gutenberg and as usual I started writing and then I went back and I wanted to add images in between some text. I often write all at once and add images later.

So I clicked on Insert in the top bar and... I noticed that hovering the blue insert bar appeared at the end of the text (btw, **wonderfully clear** interaction there). But... that's not what I need: I need to insert where my cursor is, not after the current block. I still give it a try and as hinted the new block gets added after the whole text block.

Which means that to add a new item in the middle, I had to add a new text block, copy-paste the content from the first to the second, and then add the new non-text block between the two.

**Comment by mtias** [comment]
You can only write multiple paragraphs on the classic text block. If you were using the main "text block" each paragraph would be a block, so inserting an image between two is much easier.

**Comment by folletto** [comment]
How do we explain that then? As a user they just feel the same "text".

**Comment by jasmussen** [comment]
I think _insert images at the cursor_ is a solid pattern, even if the screenshot shows the classic text block. That is — I'd love for it to work there too! 

But for the default text block, the same could work, couldn't it?

<img width="1035" alt="screen shot 2017-07-05 at 19 00 13" src="https://user-images.githubusercontent.com/1204802/27875643-441ab1dc-61b4-11e7-9b26-7aeedcbb5cb6.png">

That is —  right now it shows a blue line where the content will be inserted. But if you are in an empty text block, it should just replace it no?

**Comment by mtias** [comment]
@folletto were you editing a saved post or you explicitly inserted classic text?

**Comment by folletto** [comment]
It was a post done with an older version of Gutenberg, which is why I didn't expect the issue being the "classic text block" mode. ;) 
Or maybe it did revert back when I edited it on the old editor and back? That might have happened, but I can't recall.

Currently running 0.3.0.

Yeah I noticed now in a _new_ post that the paragraph got split automatically (lots of animations going on there tho! it covers the text of the paragraph above... should I open a ticket about that or there's one already?).

I guess it might be still a bit jarring as it's not obvious the difference between the two (apart from the bullet list icons in the quick toolbar), but I agree it's a slightly lesser issue than if it happened all the times.

> That is — right now it shows a blue line where the content will be inserted. But if you are in an empty text block, it should just replace it no?

Good point. I'm uncertain on this: once I get how Gutenberg works, I might simply not add a new line and instead insert a different block... that said, if it's empty, it might make sense to replace it. Would it be expected? What if I start a new post, I mean to write but I start by adding an image? Tough call without some user testing live.
